### PR TITLE
(HACK)Add support for restrictive selinux 

### PR DESCRIPTION
**Pull request by daviDarthemis**
In some kernels where we don't have root access, we have a break in syscalls, to solve this we add some hacks.

In kernels prior to 5.16 we do not have access to futex waitv via syscall So let's check before calling the has_waitv function

### CODE REVIEW

**Comment by Megamouse** [comment]
I can't comment on the linux code, but please fix the code style first.
It's all over the place and doesn't remotely match ours. (indentation, braces etc.)

**Review by kd-11 — COMMENTED** [review]
The changes to has_waitv are incorrect. If we call the function on older kernels, it should return ENOSYS. The function is checking for availability of the syscall which is then called elsewhere in the code. If the OS is running an older kernel it should use the fallbacks. Nothing bad happens if waitv is not supported, there are sane fallbacks in place.

The selinux flag I don't fully understand the motivation either. The application uses rare syscalls liberally, so I'd expect bypassing this one to still get others flagged. What's so special about this one specifically to warrant the hack? Wouldn't it be easier to check for both ENOSYS and EPERM as signs that the syscall is inaccessible at runtime and just report as unsupported to kick in the fallback path?

**Comment by daviDarthemis** [comment]
Indeed, the expected behavior on most Linuxes would be to simply report the lack of support, but on some systems that do not have access to advanced privileges, this may cause an error.

**Comment by daviDarthemis** [comment]
This hack removes the need for advanced privileges on some aarchs, such as Android.

As I said it's a (HACK) I'm working to solve this definitively.

However, it is enough to run games without chroot support.

Screenshot_2024-12-05-18-22-08-772_com termux x11
Screenshot_2024-12-05-14-04-26-534_com termux x11
Screenshot_2024-12-05-13-47-00-377_com termux x11

**Comment by daviDarthemis** [comment]
My next steps are:

```1• Remove this hack for a definitive version.```

```2• Adds a new, more efficient translation model in aach64 models to make better use of neon```

**Comment by kd-11** [comment]
Nothing stated in the comment is convincing enough. I'm not getting a straight answer.
1. The fallback check for older kernels is wrong, we'll still use the newer syscall in the code.
2. There is no explanation given on what happens when selinux is restricted and the syscall is invoked. Does the program crash? Or just return EPERM?
3. There's talk of a translation model but we already have native aarch64 support including neon so I don't get what you mean by this.

Either way, I see no reason to rush a hack into upstream. Take the time to polish the idea and submit when it is completed.

File: Utilities/Thread.cpp [hunk 1]
```diff
@@ -2868,6 +2872,65 @@ void thread_ctrl::detect_cpu_layout()
...
+
+void detect_big_little_topology(std::vector<int>& big_cores, std::vector<int>& little_cores, std::vector<int>& prime_cores)
+{
+    big_cores.clear();
+    little_cores.clear();
+    prime_cores.clear();
+
+    const std::string base_path = "/sys/devices/system/cpu/";
+
+    for (unsigned int i = 0; i < 8; ++i)
```
Comment (kd-11) [inline Utilities/Thread.cpp:2885]: Rpcs3 works (and is being tested by users) on asahi linux as well as ampere boards at the moment which have many more cores than this.

Comment (daviDarthemis) [reply Utilities/Thread.cpp:2885]: On arm CPU?

File: rpcs3/util/atomic.cpp [hunk 2]
```diff
@@ -29,18 +29,65 @@ namespace utils
...
+
+    static const  bool s_has_waitv = [] {
+        if (is_kernel_at_least(5, 15, 0)) {
+            // Kernel >= 5.15
+            printf("kernel >= 5.16");
+            syscall(SYS_futex_waitv, 0, 0, 0, 0, 0);
+            return errno != ENOSYS;
+        } else {
+            // Kernel < 5.15
+            printf("kernel <= 5.16");
```
Comment (Nekotekina) [inline rpcs3/util/atomic.cpp:82]: Why do you check for normal futex syscall? It's always available, nothing will work if it isn't.

Comment (kd-11) [reply rpcs3/util/atomic.cpp:82]: This whole thing looks like AI-generated code.

Comment (daviDarthemis) [reply rpcs3/util/atomic.cpp:82]: Now you're offended, I've already explained that some kernels return an error when calling syscall

**Comment by klark231** [comment]
> Nothing stated in the comment is convincing enough. I'm not getting a straight answer.
> 
> 1. The fallback check for older kernels is wrong, we'll still use the newer syscall in the code.
> 2. There is no explanation given on what happens when selinux is restricted and the syscall is invoked. Does the program crash? Or just return EPERM?
> 3. There's talk of a translation model but we already have native aarch64 support including neon so I don't get what you mean by this.
> 
> Either way, I see no reason to rush a hack into upstream. Take the time to polish the idea and submit when it is completed.

What I do know is on android we need root for RPCN. without changing selinux to permissive the network wont get through.

**Pull request closed without merging**